### PR TITLE
SUBMARINE-308. Support PyTorch with TonY runtime

### PR DESCRIPTION
**Pull request by pingsutw**
### What is this PR for?
Support PyTorch with TonY runtime

### What type of PR is it?
[Bug Fix]

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-308

### How should this be tested?
https://travis-ci.org/pingsutw/hadoop-submarine/builds/634462654

### Screenshots (if appropriate)
Screenshot from 2020-01-09 06-31-04

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by jojochuang** [comment]
Change looks good. Test failure appears unrelated.

**Comment by pingsutw** [comment]
@jojochuang thanks for the review, I re-push the patch to trigger Travis ci

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java [hunk 1]
```diff
@@ -54,22 +52,18 @@ public void setTonyClient(TonyClient client) {
...
-    if (parameters.getFramework() == Framework.PYTORCH) {
-      // we need to throw an exception, as ParametersHolder's parameters field
-      // could not be casted to TensorFlowRunJobParameters.
-      throw new UnsupportedOperationException(
-          "Support \"–-framework\" option for PyTorch in Tony is coming. " +
-              "Please check the documentation about how to submit a " +
-              "PyTorch job with TonY runtime.");
-    }
-
     LOG.info("Starting Tony runtime..");
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java]: Could you add a log entry here about the type of the framework?

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java [hunk 2]
```diff
@@ -54,22 +52,18 @@ public void setTonyClient(TonyClient client) {
...
         Constants.TONY_FINAL_XML);
     // Write user's overridden conf to an xml to be localized.
-    Configuration tonyConf = YarnUtils.tonyConfFromClientContext(
-        (TensorFlowRunJobParameters) parameters.getParameters());
+    Configuration tonyConf = null;
+    try {
+      tonyConf = YarnUtils.tonyConfFromClientContext(
+          (ParametersHolder) parameters);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to create tony conf from client context", e);
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java]: Maybe throw a `SubmarineException` here?

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 3]
```diff
@@ -42,107 +48,128 @@
...
+            parameters.getOptionValue(CliConstants.N_WORKERS));
+    if (parameters.getOptionValue(CliConstants.N_PS) != null) {
+      tonyConf.setStrings(
+              TonyConfigurationKeys.getInstancesKey(Constants.PS_JOB_NAME),
+              parameters.getOptionValue(CliConstants.N_PS));
+    }
     // Resources for PS & Worker
-    if (parameters.getPsResource() != null) {
+    if (parameters.getOptionValue(CliConstants.PS_RES) != null) {
+      Resource resource = getResource(parameters, CliConstants.PS_RES);
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]:
```suggestion
      Resource psResource = getResource(parameters, CliConstants.PS_RES);
```

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 4]
```diff
@@ -42,107 +48,128 @@
...
+              resource.getVirtualCores());
       tonyConf.setLong(
           TonyConfigurationKeys.getResourceKey(Constants.PS_JOB_NAME,
               Constants.MEMORY),
-          ResourceUtils.getMemorySize(parameters.getPsResource()));
+          ResourceUtils.getMemorySize(resource));
     }
-    if (parameters.getWorkerResource() != null) {
+    if (parameters.getOptionValue(CliConstants.WORKER_RES) != null) {
+      Resource resource = getResource(parameters, CliConstants.WORKER_RES);
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]:
```suggestion
      Resource workerResource = getResource(parameters, CliConstants.WORKER_RES);
```

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 5]
```diff
@@ -167,4 +194,14 @@ public static Configuration tonyConfFromClientContext(
 
   private YarnUtils() {
   }
+
+  private static Resource getResource(Parameter parametersHolder, String option)
+          throws ParseException, YarnException {
+    String ResourceStr =
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]:
```suggestion
    String resourceStr =
```

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 6]
```diff
@@ -167,4 +194,14 @@ public static Configuration tonyConfFromClientContext(
...
+            parametersHolder.getOptionValue(option);
+    if (ResourceStr == null) {
```
Comment (adamantal) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]:
```suggestion
    if (resourceStr == null) {
```

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java (hunk 1 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java]: Done

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java (hunk 2 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnJobSubmitter.java]: Done

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java (hunk 3 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: Done

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java (hunk 4 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: Done

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java (hunk 5 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: Done

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java (hunk 6 above)
Comment (pingsutw) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: Done

**Comment by pingsutw** [comment]
Thanks @adamantal for the review. 
I commit a new patch and add uint test for PyTorch

**Comment by pingsutw** [comment]
Will merge if no more comments

**Pull request closed without merging**